AsyncDisplayKit/ASMapNode.mm

```diff
@@ -201,33 +237,55 @@ - (void)takeSnapshot
                     // Only create a graphics context if we have annotations to draw.
                     // The MKMapSnapshotter is currently not capable of rendering annotations automatically.
                     
-                    CGRect finalImageRect = CGRectMake(0, 0, image.size.width, image.size.height);
+                    __block UIImage *pinImage;
+                    __block CGSize pinSize;
+                    __block CGPoint pinCenterOffset;
                     
-                    UIGraphicsBeginImageContextWithOptions(image.size, YES, image.scale);
-                    [image drawAtPoint:CGPointZero];
+                    void (^drawAnnotationsBlock)(void) = ^{
+                      CGRect finalImageRect = CGRectMake(0, 0, image.size.width, image.size.height);
+                          
+                      UIGraphicsBeginImageContextWithOptions(image.size, YES, image.scale);
+                      [image drawAtPoint:CGPointZero];
+                          
+                      for (id<MKAnnotation> annotation in annotations) {
+                        CGPoint point = [snapshot pointForCoordinate:annotation.coordinate];
+                        if (CGRectContainsPoint(finalImageRect, point)) {
+                          point.x -= pinSize.width / 2.0;
+                          point.y -= pinSize.height / 2.0;
+                          point.x += pinCenterOffset.x;
+                          point.y += pinCenterOffset.y;
+                          [pinImage drawAtPoint:point];
+                        }
+                      }
+                          
+                      UIImage *finalImage = UIGraphicsGetImageFromCurrentImageContext();
+                      UIGraphicsEndImageContext();
+                      strongSelf.image = finalImage;
+                    };
                     
                     // Get a standard annotation view pin. Future implementations should use a custom annotation image property.
-                    MKAnnotationView *pin = [[MKPinAnnotationView alloc] initWithAnnotation:nil reuseIdentifier:@""];
-                    UIImage *pinImage = pin.image;
-                    CGSize pinSize = pin.bounds.size;
-                    
-                    for (id<MKAnnotation> annotation in annotations) {
-                      CGPoint point = [snapshot pointForCoordinate:annotation.coordinate];
-                      if (CGRectContainsPoint(finalImageRect, point)) {
-                        CGPoint pinCenterOffset = pin.centerOffset;
-                        point.x -= pinSize.width / 2.0;
-                        point.y -= pinSize.height / 2.0;
-                        point.x += pinCenterOffset.x;
-                        point.y += pinCenterOffset.y;
-                        [pinImage drawAtPoint:point];
-                      }
+                    if (_hasCreatedPinProperties)
```

Comment:
@george-gw Good call on tracking #1890 - I think your code that calls a block could be added around here. We'd wanna make sure that it runs on the main thread, so we'd have to check the presence of the block here, and use it inside the `dispatch_async` later on.